File: daily/Sliding Window/239-Sliding-window-maximum.py
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        n = len(nums)
        max_deque = deque()
        # for i in range(k):
        #     while max_deque and nums[max_deque[-1]] < nums[i]:
        #         max_deque.pop()
        #     max_deque.append(i)
        # res.append(nums[max_deque[0]])

        for i in range(n):
            if max_deque and i - max_deque[0] >= k:
                max_deque.popleft()

            while max_deque and nums[max_deque[-1]] < nums[i]:
                   max_deque.pop()
            max_deque.append(i)
            
            if i >= k - 1:
                res.append(nums[max_deque[0]])

        # return res if k != 1 else nums
        return res

File: daily/Sliding Window/239-Sliding-window-maximum.go
package daily

func maxSlidingWindow(nums []int, k int) []int {
	res := []int{}
	n := len(nums)
	maxDeque := []int{}

	for i := 0; i < n; i++ {
		if len(maxDeque) > 0 && i-maxDeque[0] >= k {
			maxDeque = maxDeque[1:]
		}
		for len(maxDeque) > 0 && nums[maxDeque[len(maxDeque)-1]] < nums[i] {
			maxDeque = maxDeque[:len(maxDeque)-1]
		}
		maxDeque = append(maxDeque, i)
		if i >= k-1 {
			res = append(res, nums[maxDeque[0]])
		}
	}
	return res
}
